### ISSUE TITLE
string exc_value special case in Failure.__init__ misses cases

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2830]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2830|
|Type|defect|
|Created|2007-09-11 11:20:29Z|

Failure.!__init!__ has code to handle string exceptions, but only in the explicit construction case.  If a string exception is raised and a Failure is constructed like ```Failure()``` in the except suite, the check is missed (since exc_value is None there) and a Failure instance with a string for its ```type``` attribute is created.

<details><summary>Searchable metadata</summary>

```
trac-id__2830 2830
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1189509629000000 1189509629000000
changetime__1189610635000000 1189610635000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Well, it seems pretty lame that Failures constructed with explicit arguments turn their string exceptions into a thing called DefaultException and that Failures constructed with implicit arguments don't.

If I were going to implement this functionality without thinking about what's there already and backwards compatibility and such, I would probably always turn any string exception into a special type of exception.  This keeps .type and .value simple.

However, the undocumented DefaultException muddies this - is DefaultException _only_ for converted string exceptions?  I don't know.  It's also not clear whether there is code which depends on the behavior of implicitly constructed str Failures to have .type set to the actual string exception which was raised.  If I change that now, what am I breaking?

So I did something really crappy, and just made printTraceback account for the case where .type is a str.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

branch is print-string-exceptions-2830.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @exarkun|
|-|-|

Indeed, this change is not great, but.... The only thing I see is to add a comment at the change in printTraceback to remember why this check is here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Added one, how's that look?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Crystal clear, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [21280]) Merge print-string-exceptions-2830

Author: exarkun
Reviewer: therve
Fixes [#2830](https://github.com/twisted/twisted/issues/2830)

Add a test for using Failure.printTraceback on a failure created using the no-argument
initializer with a string exception.  Also add a check to Failure.printTraceback so that
it actually works in this case.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
